packages/boot/src/bootstrapper.ts

```diff
@@ -0,0 +1,127 @@
+// Node module: @loopback/boot
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Context, inject, resolveList} from '@loopback/context';
+import {
+  BootOptions,
+  BootExecutionOptions,
+  BOOTER_PHASES,
+  CoreBindings,
+  Application,
+} from '@loopback/core';
+import {resolve} from 'path';
+import {BootBindings} from './keys';
+
+import * as debugModule from 'debug';
+const debug = debugModule('loopback:boot:bootstrapper');
+
+/**
+ * The Bootstrapper class provides the `boot` function that is responsible for
+ * finding and executing the Booters in an application based on given options.
+ *
+ * NOTE: Bootstrapper should be bound as a SINGLETON so it can be cached as
+ * it does not maintain any state of it's own.
+ *
+ * @param app Appliaction instance
+ */
+export class Bootstrapper {
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) private app: Application,
+  ) {}
+
+  /**
+   * Function is responsible for calling all registered Booter classes that
+   * are bound to the Application instance. Each phase of an instance must
+   * complete before the next phase is started.
+   *
+   * @param {BootOptions} bootOptions Options for boot. Bound for Booters to
+   * receive via Dependency Injection. This normally contains config for Booters.
+   * @param {BootExecutionOptions} execOptions Execution options for boot. These
+   * determine the phases and booters that are run.
+   * @param {Context} [ctx] Optional Context to use to resolve bindings. This is
+   * primarily useful when running app.boot() again but with different settings
+   * (in particular phases) such as 'start' / 'stop'. Using a returned Context from
+   * a previous boot call allows DI to retrieve the same instances of Booters previously
+   * used as they are bound using a CONTEXT scope. This is important as Booter instances
+   * may maintain state.
+   */
+  async boot(
+    bootOptions: BootOptions,
+    execOptions?: BootExecutionOptions,
+    ctx?: Context,
+  ): Promise<Context> {
+    if (!bootOptions.projectRoot) {
+      throw new Error(
+        `No projectRoot provided for boot. Call boot({projectRoot: 'path'}) with projectRoot set.`,
+      );
+    }
+
+    const bootCtx = ctx || new Context(this.app);
+
+    // Bind booters passed in as a part of BootOptions
+    if (execOptions && execOptions.booters)
+      this.app.booter(execOptions.booters);
+
+    // Resolve path to projectRoot
+    bootOptions.projectRoot = resolve(bootOptions.projectRoot);
+
+    // Bind Boot Options for Booters
+    bootCtx.bind(BootBindings.BOOT_OPTIONS).to(bootOptions);
+
+    // Determine the phases to be run. If a user set a phases filter, those
+    // are selected otherwise we run the default phases (BOOTER_PHASES).
+    const phases = execOptions
+      ? execOptions.filter && execOptions.filter.phases
+        ? execOptions.filter.phases
+        : BOOTER_PHASES
+      : BOOTER_PHASES;
```

Comment:
Can we leave `execOptions.filter.phases` out of scope of this initial pull request please?

As I understand the currently proposed implementation, it provides rather poor user experience.
 - We have the following phases defined: `'configure', 'discover', 'load'`.
 - This API allows consumers to execute only `load` phase, skipping `configure` and `discover`. What is the expected outcome? How can we "load" artifacts when we haven't run discovery first?
 - Users can also split boot into several steps, for example call `boot()` to run "configure" & "discover" and then call boot again to run "load". Is this supported? If not, how are the users going to learn about that?

As a side note, I think that individual phases should be defined as an enum (possibly backed by a string), so that TypeScript can catch invalid phase names at compile time.

Reply:
I'm ok to drop execOptions entirely from the scope of this PR. Will discuss it with @raymondfeng.

---

But just answering some of the questions above so we have a reference for future.

The purpose for `execOptions.filter.phases` was more for advanced use cases such as those of APIC / LoopBack Workspace / advanced users? / etc. There is no check to prevent a user from running a later phase ... in which case, it's up to the `Booter` to safe guard against that scenario.

Splitting the boot into several calls to `boot()` is intentional and supported! You just need to provide the same `Context` instance returned by the `Bootstrapper.boot()` so `Booter` state is maintained across runs of `boot` with different phases. The use case of this would be introspecting an Application using the `configure, discover` phase but not `load` till the application is ready to be started. 

Another reason for `execOptions.filter.phases` was that a Booter could in theory define a new-phase such as `pre-load` and they can explain to a user how to run it by setting the appropriate filter for execOptions.